csrc/global_allocator.cpp

```diff
@@ -0,0 +1,127 @@
+              << std::endl;
+    }
+    allocated_bytes_ = 0LL;
+  }
+
+  at::Tensor getTensor(
+      const std::vector<int64_t>& sizes,
+      const c10::ScalarType& aten_dtype,
+      const c10::Device& device) {
+    // determine number of bytes needed for this tensor
+    int64_t new_bytes = dataTypeSize(aten_to_data_type(aten_dtype));
+    for (auto sz : sizes) {
+      new_bytes *= sz;
+    }
+
+    // align at 16 bytes regardless of requested dtype
+    int64_t aligned_allocated_bytes = (allocated_bytes_ + 15) & (~15);
+
+    // after this function returns this will be the allocated size
+    int64_t new_allocated_bytes = aligned_allocated_bytes + new_bytes;
+
+    // resize tensor_ if needed. Minimum size is 128B.
+    int64_t new_used_bytes = std::max((int64_t)128LL, tensor_.numel());
+    while (new_used_bytes < new_allocated_bytes) {
+      new_used_bytes *= 2;
+    }
+    if (new_used_bytes > tensor_.numel()) {
+      if (isDebugDumpEnabled(DebugDumpOption::GlobalZeroedMemory)) {
+        debug() << "[global zeroed memory] Resizing arena to " << new_used_bytes
+                << " bytes" << std::endl;
+      }
+      tensor_ = at::zeros(
+          {new_used_bytes},
+          at::TensorOptions().dtype(at::kByte).device(device));
+    }
+
+    if (isDebugDumpEnabled(DebugDumpOption::GlobalZeroedMemory)) {
+      debug() << "[global zeroed memory] Allocating byte range: "
+              << aligned_allocated_bytes << " to " << new_used_bytes << " bytes"
+              << std::endl;
+    }
+
+    // Check that memory is zeroed before allocating. Note that this launches
+    // another kernel, so it is disabled for release builds.
+#ifndef NDEBUG
+    checkZeroed();
+#endif
+
+    allocated_bytes_ = new_allocated_bytes;
+
+    // slice and view tensor
+    return tensor_
+        .index({at::indexing::Slice(
+            aligned_allocated_bytes, new_allocated_bytes, 1)})
+        .view(aten_dtype);
+  }
+
+ private:
+  void checkZeroed() const {
+    c10::Scalar nnz = at::count_nonzero(tensor_).item();
+    NVF_ERROR(
+        nnz.equal(0),
+        "Global memory arena was not properly zeroed. Found ",
+        nnz,
+        " bytes that are not zero");
+  }
+
+ private:
+  at::Tensor tensor_;
+  int64_t allocated_bytes_ = 0LL;
+};
+
+// We hold one Arena for each device
+thread_local std::vector<Arena> arenas;
+
+} // namespace
+
+at::Tensor contigZeroTensor(
+    const std::vector<int64_t>& sizes,
+    const c10::ScalarType& aten_dtype,
+    const c10::Device& device) {
+  NVF_ERROR(device.is_cuda(), "contigZeroTensor requires CUDA device");
+  int64_t device_num = device.index();
+
+  // get arena from device number, resizing arenas if needed
+  if ((size_t)device_num >= arenas.size()) {
+    arenas.resize(device_num + 1);
+  }
+
+  // request tensor from arena
+  return arenas[device_num].getTensor(sizes, aten_dtype, device);
+}
+
+void releaseZeroedMemory() {
+  for (Arena& a : arenas) {
+    a.reset();
+  }
+}
```

Comment:
A more complicated alternative would be to return a custom subclass of `at::Tensor` whose destructor marks the corresponding region as free. But given our intended usage pattern where it's always safe to release all the memory after each kernel launch, I decided to keep things simple.